Don't track unreads in archived channels

whilstsomebody opened the pull request:
Previously, messages that were unread at the time of stream archiving remained marked as unread indefinitely. Users had to manually mark them as read, which was not ideal.

With this PR, unread markers for messages are removed when the stream in which they are present is archived.

## Tasks Completed

- [x] Hide the mark as read / mark as unread options in all menus for that channel (channel, topic, and message menus).

### Hide Mark as read and unread for archived streams

 <table>
<tr>
<th>

Before

</th>
<th>

After

</th>
</tr>
<tr>
<td>

![image](https://github.com/user-attachments/assets/864af03b-730f-4e5f-b6e5-e535ab5929ec)


</td>
<td>

![image](https://github.com/user-attachments/assets/0f964955-b08b-48f0-87b8-a71b83c62e0c)

</td>
</tr>
<tr>
<td>

![image](https://github.com/user-attachments/assets/dfe22f06-7d5e-402f-8382-afebddcbe141)


</td>
<td>

![image](https://github.com/user-attachments/assets/008f345e-69ae-49a4-89be-d0094454ad5d)

</td>
</tr>
</table>

- [x] Write a database migration to clear unread markers for messages in archived channels. 

FIxes: #33259 

**I would like to thank @Aditya8840 for their help.**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

whilstsomebody committed:
stream_popover: Hide the mark as read/unread option in all menus.

When a stream is archived, we now do not show mark as read/unread
option for the messages and topics in that stream.

Aditya8840 commented on zerver/views/streams.py:
```diff
@@ -143,7 +144,14 @@ def deactivate_stream_backend(
+    if stream.recipient_id is not None:
+        for user_id in affected_user_ids:
+            affected_user_profile = UserProfile.objects.get(id=user_id)
+            do_mark_stream_messages_as_read(affected_user_profile, stream.recipient_id)
```
Have you had a chance to review the `send_subscription_remove_events` function? It might be relevant.

Aditya8840 commented on zerver/actions/streams.py:
```diff
@@ -158,6 +159,15 @@ def do_deactivate_stream(stream: Stream, *, acting_user: UserProfile | None) ->
 
+    affected_users = UserProfile.objects.filter(id__in=affected_user_ids)
+
+    send_subscription_remove_events(
```
To clarify, I'm referring specifically to [this](https://chat.zulip.org/#narrow/channel/49-development-help/topic/How.20to.20calculate.20events.20count.3F/near/2081709).

Aditya8840 commented on web/src/popover_menus_data.ts:
```diff
@@ -45,6 +45,7 @@ type ActionPopoverContext = {
     should_display_delete_option: boolean;
     should_display_read_receipts_option: boolean;
     should_display_add_reaction_option: boolean;
+    stream_not_archived: boolean;
```
Maybe we could rename this to is_stream_archived? It seems more descriptive.

whilstsomebody replied:
I think it should be `is_stream_not_archived` as this boolean is true when the stream is not archived.

whilstsomebody commented:
Hey @alya , could you please take a look at this one?

whilstsomebody commented:
Hey @alya , could you please take a look at this one?

whilstsomebody commented:
@zulipbot remove "has conflicts"

alya commented:
@sanchi-t Are you up for reviewing this one?

sanchi-t commented on web/src/stream_popover.ts:
```diff
@@ -102,12 +102,14 @@ function build_stream_popover(opts: {elt: HTMLElement; stream_id: number}): void
     const show_go_to_channel_feed =
         user_settings.web_channel_default_view !==
         web_channel_default_view_values.channel_feed.code;
+    const stream_not_archived = !sub_store.get(stream_id)?.is_archived;
```
`is_stream_archived` function already exists in `stream_data.ts`.

whilstsomebody replied:
Done!

sanchi-t commented on web/src/stream_popover.ts:
```diff
@@ -102,14 +102,14 @@ function build_stream_popover(opts: {elt: HTMLElement; stream_id: number}): void
         },
         show_go_to_channel_feed,
-        stream_not_archived,
+        is_stream_not_archived,
```
Instead of making renaming changes in a separate commit, you should update the variable name in the first commit itself.

whilstsomebody replied:
Done!

sanchi-t commented on web/templates/popovers/left_sidebar/left_sidebar_topic_actions_popover.hbs:
```diff
@@ -54,7 +54,7 @@
             </li>
             {{/if}}
-            {{#if stream_not_archived}}
+            {{#if is_stream_not_archived}}
```
These type of changes should also belong in the first commit as well.

whilstsomebody replied:
Done!

sanchi-t commented on zerver/actions/streams.py:
```diff
@@ -160,6 +160,14 @@ def do_deactivate_stream(stream: Stream, *, acting_user: UserProfile | None) ->
 
+    for user_id in affected_user_ids:
+        event = {
+            "type": "mark_stream_messages_as_read",
```
Maybe we can use `mark_stream_messages_as_read_for_everyone` instead as that was what we used to do before changing the deactivating logic. 795b2ba14e094fb13540a822cefa72fe35593841

whilstsomebody replied:
Done!

sanchi-t commented:
Also, if I have unread messages in a channel before archiving it, I end up in this state after archiving:
![image](https://github.com/user-attachments/assets/dc1a4b2e-e2a3-4b81-a180-b22275dc171b)

sanchi-t commented:
Also the keyboard shortcut works for marking messages as unread in an archived channel.

whilstsomebody committed:
unread_ops: Prevent shortcuts from marking archived messages unread.

We do not want the messages in an archived stream to get marked as
unread by shortcut keys - shift + U.

Fixes: #33259

whilstsomebody committed:
streams: Remove unread markers for messages in archived streams.

Previously, messages that were unread at the time of stream
archiving remained marked as unread indefinitely. Users had
to manually mark them as read, which was not ideal.

With this commit, unread markers for messages are removed when
the stream in which they are present is archived.

Co-authored by: Aditya Kumar Kasaudhan <74228301+Aditya8840@users.noreply.github.com>

whilstsomebody commented:
> Also, if I have unread messages in a channel before archiving it, I end up in this state after archiving: ![image](https://private-user-images.githubusercontent.com/98596642/415062999-dc1a4b2e-e2a3-4b81-a180-b22275dc171b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA1MDY0NzksIm5iZiI6MTc0MDUwNjE3OSwicGF0aCI6Ii85ODU5NjY0Mi80MTUwNjI5OTktZGMxYTRiMmUtZTJhMy00YjgxLWExODAtYjIyMjc1ZGMxNzFiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI1VDE3NTYxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdlNTllMTBhZTJhNjEzM2Y1Y2RiNmYwMmJlNWNmMTBmYzY0NmNhMGVjNWJhODJlODkzOTc3M2ZlYTI0ZDJiYTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GEJhEhpJe2_uQbh6S9YUpxOsftQ5eByt-RZ4mRnFuQY)

![image](https://github.com/user-attachments/assets/141b5ad7-6f21-4cde-91ca-03bc23108594)

Fixed this.

whilstsomebody commented:
> Also the keyboard shortcut works for marking messages as unread in an archived channel.

Fixed [here[(https://github.com/zulip/zulip/pull/33304/files#diff-a10d36cfa538ece645aca66bd12fe5bfb8e255430bc8dc8b2a76f7ab82e9029bR285-R290).

whilstsomebody commented:
Hey @sanchi-t , could you please check this again?

sanchi-t commented on web/src/stream_popover.ts:
```diff
@@ -102,12 +102,14 @@ function build_stream_popover(opts: {elt: HTMLElement; stream_id: number}): void
     const show_go_to_channel_feed =
         user_settings.web_channel_default_view !==
         web_channel_default_view_values.channel_feed.code;
+    const is_stream_not_archived = !sub_store.get(stream_id)?.is_archived;
```
You can use `is_stream_archived(stream_id: number)` function that is present in `stream_data.ts`.

sanchi-t commented on web/src/unread_ops.ts:
```diff
@@ -19,11 +19,13 @@ import type {Message} from "./message_store.ts";
 import * as recent_view_ui from "./recent_view_ui.ts";
 import type {MessageDetails} from "./server_event_types.ts";
 import type {NarrowTerm} from "./state_data.ts";
+import {is_stream_archived} from "./stream_data.ts";
```
You should `import * as stream_data` and then use `stream_data.is_stream_archived`.

sanchi-t commented on web/src/popover_menus_data.ts:
```diff
@@ -218,6 +220,8 @@ export function get_actions_popover_content_context(message_id: number): ActionP
         not_spectator &&
         !(stream_id && stream_data.is_stream_archived(stream_id));
 
+    const is_stream_not_archived = !stream_id || !stream_data.is_stream_archived(stream_id);
```
`is_stream_not_archived` is kind of not intuitive. `is_stream_archived` is probably a better name.

sanchi-t commented:
> > Also, if I have unread messages in a channel before archiving it, I end up in this state after archiving: ![image](https://private-user-images.githubusercontent.com/98596642/415062999-dc1a4b2e-e2a3-4b81-a180-b22275dc171b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA1MDY0NzksIm5iZiI6MTc0MDUwNjE3OSwicGF0aCI6Ii85ODU5NjY0Mi80MTUwNjI5OTktZGMxYTRiMmUtZTJhMy00YjgxLWExODAtYjIyMjc1ZGMxNzFiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI1VDE3NTYxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdlNTllMTBhZTJhNjEzM2Y1Y2RiNmYwMmJlNWNmMTBmYzY0NmNhMGVjNWJhODJlODkzOTc3M2ZlYTI0ZDJiYTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GEJhEhpJe2_uQbh6S9YUpxOsftQ5eByt-RZ4mRnFuQY)
> 
> ![image](https://private-user-images.githubusercontent.com/185982038/416815275-141b5ad7-6f21-4cde-91ca-03bc23108594.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA4MjMwNTksIm5iZiI6MTc0MDgyMjc1OSwicGF0aCI6Ii8xODU5ODIwMzgvNDE2ODE1Mjc1LTE0MWI1YWQ3LTZmMjEtNGNkZS05MWNhLTAzYmMyMzEwODU5NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMwMVQwOTUyMzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZDNhY2FkZTEzMjAyNWUwM2M3Y2I4YzlmMjUwMzhkYzZkM2RmNTY4Mjc4NGQxZjQxNWYzYjliMzI2YmNiYjZjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.DyUR4DmJ8uzc-Vk-hWPBRVFeQi387ojtt4rLftOQMds)
> 
> Fixed this.

I am still able to get in that state:

![image](https://github.com/user-attachments/assets/a245ea35-1093-498f-8131-80f70ac01e28)

sanchi-t reviewed (changes requested):
Also all of your commits have `Fixes: #33259` when only the last commit should have that.

whilstsomebody commented:
> > > Also, if I have unread messages in a channel before archiving it, I end up in this state after archiving: ![image](https://private-user-images.githubusercontent.com/98596642/415062999-dc1a4b2e-e2a3-4b81-a180-b22275dc171b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA1MDY0NzksIm5iZiI6MTc0MDUwNjE3OSwicGF0aCI6Ii85ODU5NjY0Mi80MTUwNjI5OTktZGMxYTRiMmUtZTJhMy00YjgxLWExODAtYjIyMjc1ZGMxNzFiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI1VDE3NTYxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdlNTllMTBhZTJhNjEzM2Y1Y2RiNmYwMmJlNWNmMTBmYzY0NmNhMGVjNWJhODJlODkzOTc3M2ZlYTI0ZDJiYTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GEJhEhpJe2_uQbh6S9YUpxOsftQ5eByt-RZ4mRnFuQY)
> > 
> > 
> > ![image](https://private-user-images.githubusercontent.com/185982038/416815275-141b5ad7-6f21-4cde-91ca-03bc23108594.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA4MjMwNTksIm5iZiI6MTc0MDgyMjc1OSwicGF0aCI6Ii8xODU5ODIwMzgvNDE2ODE1Mjc1LTE0MWI1YWQ3LTZmMjEtNGNkZS05MWNhLTAzYmMyMzEwODU5NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMwMVQwOTUyMzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZDNhY2FkZTEzMjAyNWUwM2M3Y2I4YzlmMjUwMzhkYzZkM2RmNTY4Mjc4NGQxZjQxNWYzYjliMzI2YmNiYjZjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.DyUR4DmJ8uzc-Vk-hWPBRVFeQi387ojtt4rLftOQMds)
> > Fixed this.
> 
> I am still able to get in that state:
> 
> ![image](https://private-user-images.githubusercontent.com/98596642/418267416-a245ea35-1093-498f-8131-80f70ac01e28.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA5NzY3ODIsIm5iZiI6MTc0MDk3NjQ4MiwicGF0aCI6Ii85ODU5NjY0Mi80MTgyNjc0MTYtYTI0NWVhMzUtMTA5My00OThmLTgxMzEtODBmNzBhYzAxZTI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzAzVDA0MzQ0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJjM2JiZDIwYmY5Zjk2ZGVmZmU3ZmYyZWMyODJmYThiMzY2NDc0MzE0YzZmMTkxZDhiZGZhOGFiYjAxYTNlYzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yQQjEee-N80-i6sfKGnB2_ntHvpQxfBkcjj8FE15o9c)

Fixed this.

whilstsomebody commented:
@sanchi-t , I have addressed all of your reviews. Could you please take a look again?

whilstsomebody commented:
@zulipbot remove "has conflicts"

sanchi-t commented on web/src/popover_menus_data.ts:
```diff
@@ -218,6 +220,8 @@ export function get_actions_popover_content_context(message_id: number): ActionP
         not_spectator &&
         !(stream_id && stream_data.is_stream_archived(stream_id));
 
+    const is_stream_archived = stream_id !== undefined && stream_data.is_stream_archived(stream_id);
+
```
Maybe you can calculate the value of `const is_stream_archived` variable before so that we can use the variable in the line above as well because it looks like we are applying the same logic twice.

whilstsomebody replied:
Done!

sanchi-t reviewed (commented):
Posted a minor comment. Rest looks good to me, this PR is ready for further rounds of review once you update the PR.

Also it would be nice to remove the banner once a channel is archived:

![image](https://github.com/user-attachments/assets/0fd657ce-f5e5-49df-bf5f-7a539b532c03)

whilstsomebody commented:
@sanchi-t , how can I reproduce this situation where this banner is shown. It was not displayed when I was testing otherwise I would have removed it for sure.

sanchi-t commented:
> @sanchi-t , how can I reproduce this situation where this banner is shown. It was not displayed when I was testing otherwise I would have removed it for sure?

I open a stream and mark bunch of messages as unread, then I open stream settings and archive the stream. After closing the stream settings overlay the banner is still present.

whilstsomebody commented:
> Posted a minor comment. Rest looks good to me, this PR is ready for further rounds of review once you update the PR.
> 
> Also it would be nice to remove the banner once a channel is archived:
> 
> ![image](https://private-user-images.githubusercontent.com/98596642/420004289-0fd657ce-f5e5-49df-bf5f-7a539b532c03.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEyODE1MDMsIm5iZiI6MTc0MTI4MTIwMywicGF0aCI6Ii85ODU5NjY0Mi80MjAwMDQyODktMGZkNjU3Y2UtZjVlNS00OWRmLWJmNWYtN2E1MzliNTMyYzAzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzA2VDE3MTMyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThmZjY2NmJmMTY2YjdhZjlkM2QwNjU5YTAwYjBjYzM0NWI5NDkwYjc3ZTY1MjY1Y2E5ODdjNmIyNzFkOTgwNjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XybC-B88p9uPGANQ5ToQJmrUPkFXMDx7Qg7NHkdxOBk)


@sanchi-t , this is not the a shortcoming of my PR, this is a bug in the dev environment as inspected by me. 
Reporting it on the CZO.

whilstsomebody commented:
@sanchi-t , except for the banner, I have addressed the other review.

sanchi-t commented:
@whilstsomebody LGTM!

whilstsomebody commented:
Thanks @sanchi-t for you descriptive reviews.

whilstsomebody commented:
@zulipbot drop "post release"

whilstsomebody commented:
@zulipbot add "integration review"

whilstsomebody commented:
@zulipbot remove "post release"

whilstsomebody commented:
@zulipbot add "completion candidate"
